specs/src/shwap/spec.md

```diff
@@ -0,0 +1,267 @@
+halves and the computed NMT root MUST be equal to the respective Row root in [DAH][dah].
+
+### Sample Container
+
+Sample containers encapsulate single shares of the [DataSquare][square].
+
+Sample containers are protobuf formatted using the following proto3 schema:
+```protobuf
+syntax = "proto3";
+
+message Sample {
+    bytes sample_id = 1;
+    bytes sample_share = 2;
+    Proof sample_proof = 3;
+    ProofType proof_type = 4;
+}
+
+enum ProofType {
+  RowProofType = 0;
+  ColProofType = 1;
+}
+```
+
+The fields with validity rules that form Sample containers are:
+
+[**SampleID**](#sampleid): A SampleID of the Sample container. It MUST follow [SampleID](#sampleid) formatting and field
+validity rules.
+
+**SampleShare**: A variable size array representing the share contained in the sample. Each share MUST follow [share 
+formatting and validity][shares-format] rules.
+
+**Proof**: A [protobuf formated][nmt-pb] [NMT][nmt] proof of share inclusion. It MUST follow [NMT proof verification][nmt-verify] 
+and be verified against the respective root from Row or Column axis in [DAH][dah]. The axis is defined by ProofType field.
+
+**ProofType**: An enum defining which root the Proof is coming from. It MUST be either RowProofType or ColumnProofType. 
+
+### Data Container
+
+Data containers encapsulate user submitted data under [namespaces][ns].
+
+Data containers are protobuf formatted using the following proto3 schema:
+```protobuf
+syntax = "proto3";
+
+message Data {
+    bytes data_id = 1;
+    repeated bytes data_shares = 2;
+    Proof data_proof = 3;
+}
+```
+
+The fields with validity rules that form Data containers are:
+
+[**DataID**](#dataid): A DataID of the Data container. It MUST follow [DataID](#dataid) formatting and field validity 
+rules.
+
+**DataShares**: A two-dimensional variable size byte arrays representing left data shares of a namespace in the row. 
+Each share MUST follow [share formatting and validity][shares-format] rules.
+
+**Proof**: A [protobuf formated][nmt-pb] [NMT][nmt] proof of share inclusion. It MUST follow [NMT proof verification][nmt-verify]
+and be verified against the respective root from Row or Column axis in [DAH][dah]. The axis is defined by ProofType field.
+
+Namespace data may span over multiple rows in which case all the data is encapsulated in multiple containers. This is 
+done
+
+## Protocol Extensions
+
+This section is a placeholder for future protocol extensions like new new identifiers and containers.
+
+## Considerations
+
+### Bitswap CID integration
+
+The naive question would be: "Why not to make content verification after Bitswap provided it back over its API?"
+Intuitively, this would simplify a lot and wouldn't require "hacking" CID. However, this has an important downside - 
+the Bitswap in such case would consider the content as fetched and valid, sending DONT_WANT message to its peers, while
+the message might be invalid according to the verification rules.
```

Comment:
Because the verification is done in the Bitswap level, the only way knowing that data sampling failed and block should be rejected, is to have a timeout. So, how many seconds the timeout should be? Should we define it here?

Reply:
Interesting thought. I think this timeout is out of Shwap's scope. The actual sampling logic - how, how many samples are selected, and what the timeout for sampling deserves as separate spec/cip that depends on Shwap protocol.

Reply:
Shwap is a transport for specialized DA messaging. The sampling logic/details are out of its scope